src/base/client.zig

```diff
@@ -204,10 +204,19 @@ pub fn BaseClient(comptime Reader: type, comptime Writer: type) type {
             }
         }
 
-        pub fn writeMessagePayload(self: *Self, payload: []const u8) WriterError!void {
+        pub fn writeUnmaskedPayload(self: *Self, payload: []const u8) WriterError!void {
             try self.writer.writeAll(payload);
         }
 
+        pub fn writeMessagePayload(self: *Self, payload: []const u8) WriterError!void {
+            const mask = self.current_mask.?;
+            for (payload) |c, i| {
+                try self.writer.writeByte(c ^ extractMaskByte(mask, i + self.mask_index));
+            }
+
+            self.mask_index += payload.len;
+        }
```

Comment:
We can't assume (although we technically can according to spec) that we current have a mask. This should test for `self.current_mask` and fallback to `writeUnmaskedPayload`.

Also, we should consider using our own single byte buffer and writeAll here so that it can be reused (writeByte creates a new single byte buffer on the stack on every call).

Reply:
My thinking is that if there's no mask, then this function cannot be reliably called and it's considered a dev error. Alternately, maybe return an extended errorset instead?

Reply:
Nothing about this name indicates that it would require a mask, I also believe this would be a nice simplification on the user-facing side, they can use this function if they don't want to figure out if they are sending a mask or not. At the very least we cannot use `.?` and should return an error in the case of a non-existent mask (also I think this is not preferable).

Reply:
I've been thinking this over, I think an assert is a good idea here. I would still like to get rid of the `.?` there just because it isn't very obvious what it is asserting.